ECOM 6815 Add ability to compute distinct hit and facet counts

amangano-edx opened the pull request:
This PR adds support for running elasticsearch cardinality aggregations along with search queries so that distinct hit and facet counts may be computed.

Replaces portions of https://github.com/edx/course-discovery/pull/704

@edx/ecommerce

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+            clone._distinct_hit_count = self._distinct_hit_count  # pylint: disable=protected-access
+        return clone
+
+    def set_aggregation_key(self, aggregation_key):
```
Use of getters and setters is discouraged in Python. Use a property instead.

amangano-edx replied:
Changed.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+        """
+        self._aggregation_key = aggregation_key
+
+    def get_distinct_count(self):
```
Similar to the above, use of getters and setters is discouraged in Python. You can use a `distinct_hit_count` property instead.

amangano-edx replied:
This is done this way to be consistent with the rest of the Haystack Query interface (count is exposed via `get_count` rather than a `count` property).

rlucioni replied:
@amangano-edx I see, thanks for the context. That's fair. A comment or warning to that effect (maybe at the top of the module), might make it clearer that this is largely borrowed from Haystack.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+
+        self._results = results.get('results', [])
+        self._hit_count = results.get('hits', 0)
+        self._distinct_hit_count = results.get('distinct_hits', 0)
```
In what cases would these keys be missing from the results dict?

amangano-edx replied:
Most of this is re-implemented logic from https://github.com/django-haystack/django-haystack/blob/v2.5.0/haystack/backends/__init__.py#L555. When adding custom logic, I tried to do things in a way that is consistent with how it is done in the original implementation.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+        supported_options = DistinctCountsElasticsearchBackendWrapper.SUPPORTED_FIELD_FACET_OPTIONS
+        for option, _ in options.items():
+            if option not in supported_options:
+                msg = 'DistinctCountsSearchQuery only supports a limited set of field facet options.'
```
Use parentheses when you need to join together a long string broken across multiple lines:

```py
>>> msg = (
    'this is the first line. '
    'this is the second line.'
)
>>> msg
'this is the first line. this is the second line.'
```

amangano-edx replied:
Changed.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+        """
+        Initialize a new instance of the DistinctCountsElasticsearchBackendWrapper.
+
+        -- Parameters --
```
You'll see this written as follows throughout other edX code, including this repo.
```
Arguments:
    backend (ElasticsearchSearchBackend)
    aggregation_key (string): ...
```
I'd stick to that format for consistency.

amangano-edx replied:
Changed.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+    # See the elasticsearch docs for more information about the cardinality aggregation and precision settings.
+    # https://www.elastic.co/guide/en/elasticsearch/reference/1.5/search-aggregations-metrics-cardinality-aggregation.html
+    DISTINCT_HIT_PRECISION = 1500
+    DISTINCT_FACET_PRECISION = 250
```
These seem like they belong in the settings module.

amangano-edx replied:
Changed.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+    def search(self, query_string, **kwargs):
+        """ Run a search query and return the results."""
+        if len(query_string) == 0:
+            return {'results': [], 'hits': 0, 'distinct_hits': 0}
```
What do we gain from this optimization?

amangano-edx replied:
Most of this is re-implemented logic from https://github.com/django-haystack/django-haystack/blob/v2.5.0/haystack/backends/elasticsearch_backend.py#L495. When adding custom logic, I tried to do things in a way that is consistent with how it is done in the original implementation.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+            if not self.backend.silently_fail:
+                raise
+
+            self.backend.log.error('Failed to query Elasticsearch using "%s": %s', query_string, e, exc_info=True)
```
Is `self.backend.log` an Elasticsearch construct, or is it just a Python logger under the hood? If so, you should use `log.exception`, in which case you won't need to pass `exc_info` or store the exception in `e`.

amangano-edx replied:
This is re-implemented logic from https://github.com/django-haystack/django-haystack/blob/v2.5.0/haystack/backends/elasticsearch_backend.py#L529.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+            for key in order.keys():
+                order_fields.add(key)
+
+        geo_sort = '_geo_distance' in order_fields
```
You only seem to use this variable when calling `_process_results()`. Consider doing away with the variable and determining the value inline.

amangano-edx replied:
Most of this is re-implemented logic from https://github.com/django-haystack/django-haystack/blob/v2.5.0/haystack/backends/elasticsearch_backend.py#L495. When adding custom logic, I tried to do things in a way that is consistent with how it is done in the original implementation.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+        """ Build and return a dictionary of aggregations for field facets."""
+        aggregations = {}
+        for facet_fieldname, opts in facet_dict.items():
+            for opt, _ in opts.items():
```
We typically use a double underscore for this. Single underscore is reserved for `ugettext()`.

Tangent: are the error messages in here taken from Haystack? If so, I don't think we want to mark them for translation, but if they're custom to us we might want to think about it.

amangano-edx replied:
These errors are custom. How do we mark strings for translation?

amangano-edx replied:
Changed the underscores. Will punt on the translations since these errors should only be visible to developers.

rlucioni commented on course_discovery/apps/edx_haystack_extensions/distinct_counts/backends.py:
```diff
@@ -0,0 +1,298 @@
+        for facet_fieldname, value in facet_list:
+            aggregations[facet_fieldname] = {
+                'filter': {'query': {'query_string': {'query': value}}},
+                "aggregations": self._build_cardinality_aggregation(precision=precision),
```
Minor: single-quotes for consistency, please.

amangano-edx replied:
Changed.

amangano-edx commented:
@rlucioni I believe I've addressed all of your concerns. Please take another look when you have a minute.

@mikedikan I added links to the original implementation of overridden method to their docstrings. Does that alleviate your concern?
